itest/litd_custom_channels_test.go

```diff
@@ -2046,6 +2050,166 @@ func testCustomChannelsLiquidityEdgeCases(_ context.Context,
 		"invoice, strict forwarding)")
+
+	// Edge case: Fabia gets all the asset liquidity on her side. Then
+	// generates an asset invoice to be paid for with assets by Charlie.
+	// Charlie will unleash multiple shards towards Fabia, exhausting the
+	// liquidity in the Erin<->Fabia channel. Dave should be able to detect
+	// those failures and account for those cancelled HTLCs in the RFQ
+	// tracking mechanism. Eventually we will slosh some funds from Fabia
+	// back to Erin in order for Erin to be able to forward the rest of the
+	// HTLCs, leading to the release of the preimage.
+
+	// Erin starts by sending half of the assets to Fabia's side. He also
+	// sends some raw sats to be used by Fabia later for the sloshing.
+	sendAssetKeySendPayment(
+		t.t, erin, fabia, 150_000, assetID,
+		fn.None[int64](), lnrpc.Payment_SUCCEEDED,
+		fn.None[lnrpc.PaymentFailureReason](),
+	)
+	sendKeySendPayment(t.t, erin, fabia, 20_000)
+
+	logBalance(t.t, nodes, assetID, "balance after 1st slosh")
+
+	// Fabia creates an asset invoice of 100k assets. There is currently
+	// not enough asset liquidity on Erin's side to forward HTLCs to satisfy
+	// this invoice.
+	invoiceResp = createAssetInvoice(t.t, erin, fabia, 100_000, assetID)
+
+	// We set the waiting period for the slosh payment to occur. This is set
+	// to half of the default payment timeout, as we want it to occur half
+	// way through the in-flight payment.
+	sloshWait := PaymentTimeout / 3
+
+	go func() {
+		// After a small delay (less than the payment timeout) Fabia
+		// sloshes back the asset liquidity to Erin. This should allow
+		// the payment by Charlie to eventually complete.
+		time.Sleep(sloshWait)
+		sendAssetKeySendPayment(
+			t.t, fabia, erin, 75_000, assetID,
+			fn.None[int64](), lnrpc.Payment_SUCCEEDED,
+			fn.None[lnrpc.PaymentFailureReason](),
+		)
+	}()
+
+	// To avoid goroutine uncertainty, we wait as much as the above routine
+	// minus a small delta. This is enough for us to be sure that the
+	// payment will complete due to the slosh.
+	timeoutChan = time.After(sloshWait - time.Millisecond*250)
+	done = make(chan bool, 1)
+
+	go func() {
+		// Now Charlie pays the invoice with assets. What happens on the
+		// Charlie-Dave channel doesn't matter. This payment will block
+		// until the previous slosh payment completes. That slosh will
+		// allow for the rest of the HTLCs to be forwarded.
+		payInvoiceWithAssets(
+			t.t, charlie, dave, invoiceResp.PaymentRequest, assetID,
+			true, defaultPaymentStatusOpt,
+		)
+
+		done <- true
+	}()
+
+	select {
+	case <-done:
+		// If the payment completes before the slosh payment occurs then
+		// something went wrong, this is not the expected test case
+		// behavior.
+		t.Fatalf("payment completed before expected block period")
+	case <-timeoutChan:
+		// The expected time delay for the payment to complete has been
+		// passed, so now we wait for the payment to complete.
+		<-done
+	}
```

Comment:
We have `BeforeTimeout` as a helper function. Would be good to have `AfterTimestamp` perhaps. But I think this sort of timing analysis strategy should be events driven instead.

Reply:
thanks, will look into `BeforeTimeout` and see if it can save us some lines

Reply:
not needed anymore with new approach